iroh-rpc/src/stream.rs

```diff
@@ -1,191 +1,61 @@
-    ) {
-        loop {
-            select! {
-              res = packet_receiver.next() => {
-                if let Some(r) = res {
-                  match r {
-                    StreamType::Packet(p) => {
-                      {
-                        let mut chunks = chunks.lock().expect("failed to lock stream chunk mutex");
-                        chunks.insert(p.index, p.data);
-                      }
-                      debug!(target="OrderedStream", "Stream received chunk {}", p.index);
-                  },
-                    StreamType::RpcError(e) => {
-                      error!(target="OrderedStream", "stream received error {}", e);
-                      error_sender
-                          .send(e)
-                          .await
-                          .expect("failed to send on error sender");
-                    },
-                  }
-                  // notify poll_next that we have a message we want them to
-                  // process
-                  {
-                    let task = task.lock().expect("failed to lock task");
-                    task.wake();
-                  }
-                  continue;
-                }
-                // if res == None, then the packet_receiver stream has closed
-                    debug!(target="OrderedStream", "Packet receiver closed, done gathering packets");
-                      return;
-                                },
-              _ = shutdown => {
-                debug!(target="OrderedStream", "Shutdown triggered, no longer gathering packets");
-                return;
-              }
-            }
-        }
-    }
-
-    /// close cleans up the stream. MUST BE CALLED AFTER YOU ARE FINISHED WITH THE STREAM.
-    pub async fn close(mut self) {
-        self.out_sender
-            .try_send(Command::CloseStream { id: self.id })
-            .expect("Out sender to still be active.");
-        if let Some(sender) = self.early_shutdown.take() {
-            debug!(target = "OrderedStream", "closing stream");
-            let _ = sender
-                .send(())
-                .expect("shutdown receiver should still be open");
-            return;
-        } else {
-            debug!(target = "OrderedStream", "stream already closed");
-        }
-        match self.handle.await {
-            Ok(()) => {
-                debug!("stream closed gracefully");
+pub fn make_order(
+    packet_receiver: StreamReceiver,
+) -> impl Stream<Item = Result<Vec<u8>, RpcError>> {
+    let max = 1024;
+    let mut chunks = HashMap::with_capacity(max);
+    let mut next_chunk = 0;
+
+    stream! {
+        for await packet in packet_receiver {
+            match packet {
+                StreamType::Packet(p) => {
+                    if next_chunk == p.index {
+                        next_chunk += 1;
+                        yield Ok(p.data);
+                    } else {
+                        if let Some(chunk) = chunks.remove(&next_chunk) {
```

Comment:
this should yield all elements that are available, not just one

Reply:
damn i really underestimated the power of `yield`